
# Step 1. reate a .geno file for input to the painting script. 

# this file is based of the final filtered VCF. To getnerate the file I follow Simon Martin's script. 
# To use these there a depedencies that have conda installations. As conda is depreciated on puhti
# these conda envrionments need to first be containerized with tykky.

# clone simon martin's git repositry

git clone https://github.com/simonhmartin/genomics_general.git

# running the code 
# need to make a tykky container for the conda environment
# on local machine:
conda create -n painting_env
source activate painting_env
conda install -c etetoolkit ete3 numpy pandas
conda env export --from-history -n painting_env | grep -v prefix > painting_env.yml

# on puhti

conda-containerize new --prefix <install_dir> painting_env.yml

### ------------ DONE CONTAINERIZING --------------------------


# Step 2. Remove the unused samples from the VCF

# the analysis only used male samples to comply with the Li et al methods requirement of only homozygous sites
# i.e. no AIMs will need to be dropped due to being heterozygous in females.



module load biokit

## filter VCf to only male samples
cd /scratch/project_2003480/patrick/data/vcf/filtered

VCF=final_samples.normalized.SnpGap_2.NonSNP.Balance.PASS.decomposed.SNPQ30.biall.fixedHeader.indDP.hwe.MminDP3FminDP6.indMiss45.siteMiss40.vcf.gz
MALES=male_samples.list

bcftools view -S $MALES $VCF | head
bcftools view -S $MALES -Ou $VCF | bcftools view -g hom - | bgzip > final_males_forAD.normalized.SnpGap_2.NonSNP.Balance.PASS.decomposed.SNPQ30.biall.fixedHeader.indDP.hwe.MminDP3FminDP6.indMiss45.siteMiss40.vcf.gz


## ------ DONE PREPPING VCF -----------------------------------

# Step 3. Make ploidy file to parseVCF.py (DEPRECIATED) and sample name list

# The script for making a .geno files has a ploidy file as input. it is a two column tab separated file
# with sample name os one column and ploidy as the second column.


## make ploidy file

cd /scratch/project_2003480/patrick/analyses/ADscan

# N_MALES=$(wc -l /scratch/project_2003480/patrick/data/vcf/filtered/male_samples.list | cut -f 1 -d ' ')
# for i in $(seq 1 $N_MALES); do echo "1" >> ploidy.list; done

# paste -d '\t' male_samples.list ploidy.list > /scratch/project_2003480/patrick/analyses/ADscan/ploidy.tsv

# ## make a comma speparated list of smaple names

for i in $(cat /scratch/project_2003480/patrick/data/vcf/filtered/male_samples.list)
do 
echo -n "${i},"
done

#m001-2005,m001-2011,m001-2018,m002-2005,m002-2011,m002-2018,m003-2005,m003-2011,m003-2018,m004-2005,m004-2018,m005-2005,m005-2011,m005-2018,m006-2005,m006-2011,m006-2014,m006-2018,m007-2005,m007-2011,m007-2014,m008-2005,m008-2011,m009-2005,m009-2011,m009-2018,m010-2005,m010-2011,m011-2005,m012-2005,m012-2011,m012-2014,m012-2018,m013-2005,m013-2021,m014-2005,m014-2011,m014-2014,m014-2018,m015-2005,m015-2014,m015-2018,m016-2005,m016-2014,m016-2018,m017-2005,m017-2014,m017-2018,m018-2005,m018-2014,m018-2018,m019-2005,m019-2018,m020-2005,m020-2014,m020-2021,m021-2005,m021-2014,m021-2018,m021-2021,m022-2005,m022-2018,m022-2020,m022-2021,m023-2005,m023-2014,m023-2020,m023-2021,m024-2005,m024-2014,m024-2018,m024-2020,m024-2021,m025-2004,m025-2005,m025-2014,m025-2018,m025-2020,m025-2021,m026-2004,m026-2005,m026-2014,m026-2018,m026-2021,m027-2004,m027-2005,m027-2020,m027-2021,m028-2004,m028-2005,m028-2018,m028-2020,m028-2021,m029-2004,m029-2005,m029-2018,m029-2020,m029-2021,m030-2004,m030-2005,m030-2014,m030-2018,m030-2020,m030-2021,m031-2004,m031-2005,m031-2014,m031-2018,m031-2020,m031-2021,m032-2004,m032-2005,m032-2014,m032-2018,m033-2004,m033-2005,m033-2014,m033-2018,m033-2020,m033-2021,m034-2005,m034-2014,m034-2018,m034-2020,m035-2004,m035-2005,m035-2014,m035-2018,m035-2020,m035-2021,m036-2005,m036-2014,m036-2018,m036-2020,m036-2021,m037-2004,m037-2005,m037-2014,m037-2018,m037-2020,m038-2004,m038-2005,m038-2014,m038-2018,m038-2021,m039-2004,m039-2005,m039-2014,m039-2018,m039-2020,m040-2004,m040-2005,m040-2014,m040-2018,m040-2020,m040-2021,m041-2004,m041-2018,m041-2020,m041-2021,m042-2004,m042-2014,m042-2020,m042-2021,m043-2004,m043-2014,m043-2020,m043-2021,m044-2004,m044-2020,m044-2021,m045-2014,m045-2018,m045-2020,m045-2021,m046-2004,m046-2014,m046-2020,m046-2021,m047-2014,m047-2020,m047-2021,m048-2004,m048-2020,m048-2021,m049-2004,m049-2020,m049-2021,m050-2014,m050-2020,m050-2021,m051-2004,m051-2014,m051-2018,m051-2020,m051-2021,m052-2020,m052-2021,m053-2014,m053-2020,m053-2021,m054-2020,m055-2020,m055-2021,m056-2020,m056-2021,m057-2004,m057-2020,m057-2021,m058-2020,m058-2021,m059-2004,m059-2020,m059-2021,m060-2004,m060-2018,m060-2020,m061-2018,m061-2020,m062-2018,m063-2004,m063-2018,m064-2004,m064-2018,m065-2004,m065-2018,m066-2004,m066-2018,m068-2004,m069-2004,m069-2018,m070-2004,m070-2018,m071-2004,m072-2004,m073-2004,m073-2018,m074-2018,m075-2018,m076-2004,m077-2004,m078-2004,m079-2004,m079-2018,m080-2004,m081-2004,

## Step 4. Generate the .geno file

# this file will serve as the input to the painting script along with the AIMs file.

# From this point the exports are just run first.

cd /scratch/project_2003480/patrick/analyses/ADscan

export PATH="/projappl/project_2003480/painting_env/bin:$PATH"

export PYTHONPATH="$PROJAPPL/genomics_general:$PYTHONPATH" # don't forget this step or it won't work

module load biokit

module load biopythontools/11.3.0_3.10.6_1.76

VCF=/scratch/project_2003480/patrick/data/vcf/filtered/final_males_forAD.normalized.SnpGap_2.NonSNP.Balance.PASS.decomposed.SNPQ30.biall.fixedHeader.indDP.hwe.MminDP3FminDP6.indMiss45.siteMiss40.vcf.gz
#PLOIDY=/scratch/project_2003480/patrick/analyses/ADscan/ploidy.tsv # should be file with two columnd sample name and ploidy as 1s

# python $PROJAPPL/genomics_general/VCF_processing/parseVCF.py \
#   -i $VCF \
#   --skipIndels \
#   --ploidyFile $PLOIDY \
#   -s m001-2005,m001-2011,m001-2018,m002-2005,m002-2011,m002-2018,m003-2005,m003-2011,m003-2018,m004-2005,m004-2018,m005-2005,m005-2011,m005-2018,m006-2005,m006-2011,m006-2014,m006-2018,m007-2005,m007-2011,m007-2014,m008-2005,m008-2011,m009-2005,m009-2011,m009-2018,m010-2005,m010-2011,m011-2005,m012-2005,m012-2011,m012-2014,m012-2018,m013-2005,m013-2021,m014-2005,m014-2011,m014-2014,m014-2018,m015-2005,m015-2014,m015-2018,m016-2005,m016-2014,m016-2018,m017-2005,m017-2014,m017-2018,m018-2005,m018-2014,m018-2018,m019-2005,m019-2018,m020-2005,m020-2014,m020-2021,m021-2005,m021-2014,m021-2018,m021-2021,m022-2005,m022-2018,m022-2020,m022-2021,m023-2005,m023-2014,m023-2020,m023-2021,m024-2005,m024-2014,m024-2018,m024-2020,m024-2021,m025-2004,m025-2005,m025-2014,m025-2018,m025-2020,m025-2021,m026-2004,m026-2005,m026-2014,m026-2018,m026-2021,m027-2004,m027-2005,m027-2020,m027-2021,m028-2004,m028-2005,m028-2018,m028-2020,m028-2021,m029-2004,m029-2005,m029-2018,m029-2020,m029-2021,m030-2004,m030-2005,m030-2014,m030-2018,m030-2020,m030-2021,m031-2004,m031-2005,m031-2014,m031-2018,m031-2020,m031-2021,m032-2004,m032-2005,m032-2014,m032-2018,m033-2004,m033-2005,m033-2014,m033-2018,m033-2020,m033-2021,m034-2005,m034-2014,m034-2018,m034-2020,m035-2004,m035-2005,m035-2014,m035-2018,m035-2020,m035-2021,m036-2005,m036-2014,m036-2018,m036-2020,m036-2021,m037-2004,m037-2005,m037-2014,m037-2018,m037-2020,m038-2004,m038-2005,m038-2014,m038-2018,m038-2021,m039-2004,m039-2005,m039-2014,m039-2018,m039-2020,m040-2004,m040-2005,m040-2014,m040-2018,m040-2020,m040-2021,m041-2004,m041-2018,m041-2020,m041-2021,m042-2004,m042-2014,m042-2020,m042-2021,m043-2004,m043-2014,m043-2020,m043-2021,m044-2004,m044-2020,m044-2021,m045-2014,m045-2018,m045-2020,m045-2021,m046-2004,m046-2014,m046-2020,m046-2021,m047-2014,m047-2020,m047-2021,m048-2004,m048-2020,m048-2021,m049-2004,m049-2020,m049-2021,m050-2014,m050-2020,m050-2021,m051-2004,m051-2014,m051-2018,m051-2020,m051-2021,m052-2020,m052-2021,m053-2014,m053-2020,m053-2021,m054-2020,m055-2020,m055-2021,m056-2020,m056-2021,m057-2004,m057-2020,m057-2021,m058-2020,m058-2021,m059-2004,m059-2020,m059-2021,m060-2004,m060-2018,m060-2020,m061-2018,m061-2020,m062-2018,m063-2004,m063-2018,m064-2004,m064-2018,m065-2004,m065-2018,m066-2004,m066-2018,m068-2004,m069-2004,m069-2018,m070-2004,m070-2018,m071-2004,m072-2004,m073-2004,m073-2018,m074-2018,m075-2018,m076-2004,m077-2004,m078-2004,m079-2004,m079-2018,m080-2004,m081-2004 | bgzip > your.geno.gz

# raise ValueError("Sample {} at {}:{} genotype {} does not match explected ploidy of {}".format(sample, self.CHROM, self.POS,
# ValueError: Sample m001-2005 at Scaffold01:6385 genotype ./. does not match explected ploidy of 1 

# error is that male genotypes in the vcf are already coded as diploid, jsut all homozygous


python $PROJAPPL/genomics_general/VCF_processing/parseVCF.py \
  -i $VCF \
  --skipIndels \
  -s m001-2005,m001-2011,m001-2018,m002-2005,m002-2011,m002-2018,m003-2005,m003-2011,m003-2018,m004-2005,m004-2018,m005-2005,m005-2011,m005-2018,m006-2005,m006-2011,m006-2014,m006-2018,m007-2005,m007-2011,m007-2014,m008-2005,m008-2011,m009-2005,m009-2011,m009-2018,m010-2005,m010-2011,m011-2005,m012-2005,m012-2011,m012-2014,m012-2018,m013-2005,m013-2021,m014-2005,m014-2011,m014-2014,m014-2018,m015-2005,m015-2014,m015-2018,m016-2005,m016-2014,m016-2018,m017-2005,m017-2014,m017-2018,m018-2005,m018-2014,m018-2018,m019-2005,m019-2018,m020-2005,m020-2014,m020-2021,m021-2005,m021-2014,m021-2018,m021-2021,m022-2005,m022-2018,m022-2020,m022-2021,m023-2005,m023-2014,m023-2020,m023-2021,m024-2005,m024-2014,m024-2018,m024-2020,m024-2021,m025-2004,m025-2005,m025-2014,m025-2018,m025-2020,m025-2021,m026-2004,m026-2005,m026-2014,m026-2018,m026-2021,m027-2004,m027-2005,m027-2020,m027-2021,m028-2004,m028-2005,m028-2018,m028-2020,m028-2021,m029-2004,m029-2005,m029-2018,m029-2020,m029-2021,m030-2004,m030-2005,m030-2014,m030-2018,m030-2020,m030-2021,m031-2004,m031-2005,m031-2014,m031-2018,m031-2020,m031-2021,m032-2004,m032-2005,m032-2014,m032-2018,m033-2004,m033-2005,m033-2014,m033-2018,m033-2020,m033-2021,m034-2005,m034-2014,m034-2018,m034-2020,m035-2004,m035-2005,m035-2014,m035-2018,m035-2020,m035-2021,m036-2005,m036-2014,m036-2018,m036-2020,m036-2021,m037-2004,m037-2005,m037-2014,m037-2018,m037-2020,m038-2004,m038-2005,m038-2014,m038-2018,m038-2021,m039-2004,m039-2005,m039-2014,m039-2018,m039-2020,m040-2004,m040-2005,m040-2014,m040-2018,m040-2020,m040-2021,m041-2004,m041-2018,m041-2020,m041-2021,m042-2004,m042-2014,m042-2020,m042-2021,m043-2004,m043-2014,m043-2020,m043-2021,m044-2004,m044-2020,m044-2021,m045-2014,m045-2018,m045-2020,m045-2021,m046-2004,m046-2014,m046-2020,m046-2021,m047-2014,m047-2020,m047-2021,m048-2004,m048-2020,m048-2021,m049-2004,m049-2020,m049-2021,m050-2014,m050-2020,m050-2021,m051-2004,m051-2014,m051-2018,m051-2020,m051-2021,m052-2020,m052-2021,m053-2014,m053-2020,m053-2021,m054-2020,m055-2020,m055-2021,m056-2020,m056-2021,m057-2004,m057-2020,m057-2021,m058-2020,m058-2021,m059-2004,m059-2020,m059-2021,m060-2004,m060-2018,m060-2020,m061-2018,m061-2020,m062-2018,m063-2004,m063-2018,m064-2004,m064-2018,m065-2004,m065-2018,m066-2004,m066-2018,m068-2004,m069-2004,m069-2018,m070-2004,m070-2018,m071-2004,m072-2004,m073-2004,m073-2018,m074-2018,m075-2018,m076-2004,m077-2004,m078-2004,m079-2004,m079-2018,m080-2004,m081-2004 | bgzip > your.geno.gz
